src/rust/derive-dynamic-node/src/lib.rs

```diff
@@ -288,10 +345,117 @@ fn property_methods(property_name: &Ident, property_type: &Type) -> TS2 {
     implementation
 }
 
-#[cfg(test)]
-mod tests {
-    #[test]
-    fn it_works() {
-        assert_eq!(2 + 2, 4);
+#[proc_macro_derive(GraplSessionId, attributes(grapl))]
+pub fn derive_grapl_session(input: TokenStream) -> TokenStream {
+    let input: syn::DeriveInput = syn::parse_macro_input!(input as syn::DeriveInput);
+
+    let input_struct = match input.data {
+        Data::Struct(input_struct) => input_struct,
+        _ => panic!("Only available for struct"),
+    };
+
+    let fields = match input_struct.fields {
+        Fields::Named(fields) => fields.named,
+        _ => panic!("Requires named fields"),
+    };
+
+    let mut created_time_prop: Option<String> = None;
+    let mut last_seen_time_prop: Option<String> = None;
+    let mut terminated_time_prop: Option<String> = None;
+
+    for field in fields.iter() {
+        set_timestamp_from_meta(field, CREATE_TIME, &mut created_time_prop);
+        set_timestamp_from_meta(field, LAST_SEEN_TIME, &mut last_seen_time_prop);
+        set_timestamp_from_meta(field, TERMINATE_TIME, &mut terminated_time_prop);
+    }
+
+    if created_time_prop.is_none() {
+        panic!("Must set created_time for at least one field");
+    }
+
+    if last_seen_time_prop.is_none() {
+        panic!("Must set last_seen_time for at least one field");
+    }
+
+    if terminated_time_prop.is_none() {
+        panic!("Must set terminated_time for at least one field");
+    }
+    let mut id_fields = quote!();
+    for field in fields {
+        for attr in &field.attrs {
+            on_grapl_attrs(attr, |meta_attr| {
+                if meta_attr == "pseudo_key" {
+                    let f = field
+                        .ident
+                        .as_ref()
+                        .expect("field is missing an identifier")
+                        .to_string();
+                    id_fields.extend(quote!(#f .to_string(), ));
+                }
+            });
+        }
+    }
+
+    let struct_name = &input.ident;
+
+    let node_name_str = format!("{}Node", struct_name);
+    let node_name = syn::Ident::new(&node_name_str, struct_name.span());
+    // Add node name to id
+    let q = quote!(
+        impl #node_name {
+            pub fn session_strategy() -> IdStrategy {
```

Comment:
should these 2 functions be a trait interface?  `impl #node_name for StrategicNode` or w/e

Reply:
They could be made part of a separate trait but I don't think there's a need in this case. The reason why we use a trait for the other methods is so that we can override the trait impl if necessary, but given that this is a single static method you could just as easily write it out yourself if you wanted to override it.

Reply:
it's a single static method, but you can never call it generically, because there is no static conceptualization of "any node with this given session strategy".  you just have a bunch of functions that coincidentally have the same name.  this makes it impossible to directly test, or implement default functions for dynamic nodes outside of this macro code.

Reply:
Gotcha, yeah I think that's reasonable but I'd like that to be a separate PR.